packages/lit-dev-content/samples/examples/react-events/index.tsx

```diff
@@ -0,0 +1,27 @@
+import {createComponent} from "@lit-labs/react";
+
+import {React, ReactDOM, CounterButton as CounterButtonWC} from "./deps.js";
+
+const CounterButton = createComponent(
+    React,
+    'counter-button',
+    CounterButtonWC,
+);
+  
+const App = () => {
+    const [count, setCount] = React.useState(0);
+
+    const clickCallback = React.useCallback((e) => {
+        setCount(count + 1);
+    }, [count, setCount])
```

Comment:
this is incorrect. it will only ever update once.

Reply:
but changing the dependency to `[count]` would simplify the dependency chain.

We need to generate a new `callback` function every time `count`  updates because it retains the scope of `count` and immediately becomes stale after a render.

Reply:
The point of using `useCallback` would be to not have it re-declare the function every time this component re-renders via state change. If you're wanting to re-declare on `count`, then it would be simpler to just declare a handler function without `useCallback`, or do one inline like `onClick={() => setCount(count + 1)}`.

My suggestion works with just a single function declaration via `useCallback` that can be reused because it provides a function that gets provided the previous (or current depending on your point of view) value of the state.

Reply:
Sorry if I'm harping on too much about this but I'm just really passionate about React hooks stuff as you are ❤️.

This is definitely nit-picking territory in terms of whether this is necessary for small examples like this, but if we're talking "best practices", I think putting some thoughts in is worth it.

In the code below...
```js
// will create an anonymous function every click ....
  const clickCallback = React.useCallback((e) => {
        setCount((prevCount) => prevCount + 1);
                 ^^^^^^^ here
  }, [setCount])
```
The `(prevCount) => prevCount + 1` function is part of the function definition of the arrow function `(e) => ...` passed into `useCallback()` which only gets re-declared when the dependency changes, not on every click. Moving that out to a named function *outside* of the component function scope could save on that being re-declared but I don't know if it's worth all that.

Trying to avoid the anonymous functions isn't exactly the priority here. The original implementation with `useCallback` already creates a new anonymous function every time a button is clicked.
```js
    const clickCallback = React.useCallback((e) => {
                                            ^^^ here
        setCount(count + 1);
    }, [count, setCount])
```
But with `count` in the dependency array, `clickCallback` will also be a new function reference every time the button is clicked and the `count` changes, which means the component that gets `clickCallback` as prop will re-render. So in my eyes, that's no better than `onClick={() => setCount(count + 1)}` but has the extra overhead of `useCallback()` call.

You are absolutely correct that `setCount` is stable and always the same reference so can be omitted from the dependency array. I think at some point the lint rule of `exhaustive-deps` got updated to not complain about not including those in the dependency array.

Function update for `setState` is not a bad practice and, if anything, preferable to use when the update depends on the previous value of the state as referencing the value of `count` in scope is unreliable like when you click the button multiple times really quickly before the state updates or if `setState` is called after some time asynchronously.
https://reactjs.org/docs/hooks-reference.html#functional-updates
https://stackoverflow.com/a/57828519/19498660
https://betterprogramming.pub/you-dont-know-usestate-until-you-ve-used-functional-updates-5da52117620f

So doing this...
```js
    const clickCallback = React.useCallback((e) => {
        setCount((prevCount) => prevCount + 1);
    }, []);
```
or
```js
const setter = (prevCount) => prevCount + 1;

const App = () => {
    // ...
    const clickCallback = React.useCallback((e) => {
        setCount(setter);
    }, []);
}
```
`clickCallback` will be a stable reference across clicks and any component that gets it as a prop doesn't have to re-render because of it.

Reply:
Hmm. Yes, the setter passed into the `setCount` function does get re-declared every time the button is clicked, i.e. when the callback is called. This can be avoided by moving that `setter` outside of component scope as mentioned. But that extra function generation is not what I'm concerned about.

Here's a somewhat contrived example in stackblitz.
https://stackblitz.com/edit/react-ts-gim2vk?file=App.tsx

With your approach (A), the `Button` component has to re-render every time it is clicked and has to re-register the event handler whereas in mine (B) it can be memoized because the `handleChange` prop is always the same reference across clicks. This can be seen by looking at the logs in the console.

I've also added async versions of these where it takes a second before the count updates.
In A, if you click the async button multiple times, the count only increments by 1. If you click the async button once then click the sync button multiple times within the second, the count turns to one more than what it was when the async button was clicked, ignoring all the sync clicks in between.
In B, every button click, sync or async, results in an extra count. Multiple async clicks will queue up and cause an increment each, and adding sync clicks in between won't affect it.

It probably depends on the real use case which behavior is preferable but, in general, B feels more right to me.

Reply:
Oh I missed the second comment.

```js
const clickCallback = React.useCallback((e) => {
      setCount((prevCount) => prevCount + 1);
               ^^^^^^^ here
}, [setCount])
```
That anonymous function of `(prevCount) => prevCount + 1` is only created when the outer  function `(e) => { ... }` that contains it is called and evaluated. Only the memoized version of this function is called on each click. So it is wrong to say that that particular anonymous function `(prevCount) => ...` is generated and never unused on every render. It is the `(e) => { ... }` that is generated and never used. And this is generated every render in both approach.

Reply:
Yes I agree partially and mentioned this in previous comments. The anonymous function will be created every click.

>The following [your example] creates an anonymous callback function which is never used every render (the anonymous callback). However the initial callback created (first render) will always generate new anonymous functions that are created and used with every click.
 
 
And in the first example in my first comment
```
// will create an anonymous function every click ....
  const clickCallback = React.useCallback((e) => {
        setCount((prevCount) => prevCount + 1);
                 ^^^^^^^ here
  }, [setCount])
  
...

// and every render ...
onClick={() => setCount(count + 1)}
         ^^^^^^ here
```

Reply:
I feel like your async example highlights the `scope` issues i brought up earlier.

```
  const handleClick = React.useCallback(() => {
    setCount(count + 1);
  }, [count]);

  const handleClickAsync = React.useCallback(() => {
    sleep(1000).then(() => {
      setCount(count + 1);
    });
  }, [count]);
```

The reason those number don't add up is becauce the scope of the async function is stale. I also would avoid async anything with hooks for this reason.

The correct way to handle this is to use a reducer and have async functions send an `intention` or `action`. I'll make an example shortly.

Reply:
Why add extra complexity with a reducer when `setState` already comes with the ability to perform an action based on the current value of the `state` regardless of when it's called?

Why re-declare a callback that does the same job -- take previous value and add one to it -- every time the state changes just to update the scope that gets stale?

It's because of that scoping issue that `setState` can take a function that will be provided the latest value.

Reply:
Official React doc: https://reactjs.org/docs/hooks-reference.html#functional-updates
Official beta React doc: https://beta.reactjs.org/apis/usestate#updating-state-based-on-the-previous-state

I don't buy that this approach is "complex" because it's not mentioned in the "Intro to Hooks" section of the old doc.

As for using `useReducer`, you're now complicating the example to justify its use. Of course if you're handling more than 1 piece of independent state, going with `useReducer` starts making sense. If you're arguing `useState` doesn't scale, are you suggesting you should never use `useState` and always use `useReducer`? Not every situation requires that.

If the using the updater function really bothers you, then sure we don't need to use it. Like the beta doc mentions, in most cases, it's not necessary. However, I brought it up because you're also using `useCallback`.

```js
  const [count, setCount] = React.useState(0);

  const clickCallback = React.useCallback((e) => setCount(count + 1), [count]);
```

This code is nonsensical to me because there's no point in using `useCallback` since the only time this component would re-render is when `count` changes which also invalidates the memoization.

It would then be just better to leave it as

```js
  const [count, setCount] = React.useState(0);

  const clickCallback = (e) => setCount(count + 1);
```

Same effect, no overhead of `useCallback`.

If you want to use `useCallback` though, keeping the dependency array empty by using the updater function in the setState would be the way to go. The reason for using `useCallback` is to take advantage of the memoization, keeping the reference stored in `clickCallback` the same across re-renders as it's being passed to child components. Admittedly this kind of performance squeezing is not really necessary for this use case, and we won't be taking full advantage of that reference stability without also reaching for `React.memo` which is certainly overkill for an example like this. So if we're going for simplicity, might as well go with that last code snippet, or even just provide the handler as an anonymous arrow function. That is, after all, how React's own doc does it.

Reply:
We should talk in person. This isn't going anywhere and I don't want to put a "cool trick" into a simple example.

Reply:
> As for using useReducer, you're now complicating the example to justify its use.

I'm not creating a more "complex" example. I'm using your examples to explain the difference between our approaches.

They also highlight where your pattern breaks down, which is when any amount of complexity is added.

Please look at the examples again

> So if we're going for simplicity, might as well go with that last code snippet

I don't consider a function in use state simple. I would not let that pass through a code review and I wouldn't teach someone that.

> just provide the handler as an anonymous arrow function

I'd rather do that than use a function in a `useState`. It's clearer and easier to read.

Personally I don't understand why we are stuck on something so standard.

I've done my best to explain my reasoning here. Would be more than happy to explain in person

Reply:
Looking back on this thread, I feel like I've explained why I don't want to use an `updater` from a technical, professional, and academic perspective. (as well as a personal one)

If you'd like to submit a PR later that adds `updaters` to these examples that's totally fine. I don't want that pattern associated with my submissions.


Looking through the React beta docs, they explicitly state the same limitations I've been trying to point out.

### extra steps (purity checks)

>React may [call your updaters twice](https://beta.reactjs.org/apis/usestate#my-initializer-or-updater-function-runs-twice) in development to verify that they are [pure](https://beta.reactjs.org/learn/keeping-components-pure).

It is an extra effort for react to use `updaters`, it is not exactly the standard use case

### difficult to extend

Reacts beta docs highlights how limiting this approach actually is:
Notice the +3 vs +1 callback, that simply uses the setState x3 vs x1

That’s where that pattern ends, it’s not extensible.

Here's reacts own code example for `updaters`
```js
export default function Counter() {
  const [age, setAge] = useState(42);

  function increment() {
    setAge(a => a + 1);
  }

  return (
    <>
      <h1>Your age: {age}</h1>
      <button onClick={() => {
        increment();
        increment();
        increment();
      }}>+3</button>
      <button onClick={() => {
        increment();
      }}>+1</button>
    </>
  );
}
```
I wouldn't exactly call this well written or well made. There’s no way to make this dynamic without it falling apart. To me that means it’s not a durable foundation for users to build upon

And that's even before we start to critique this example for quick optimizations. It's far from optimal.

### Stateful issues

>  If it’s calculated from the previous state of some _other_ state variable, you might want to combine them into one object and [use a reducer](https://beta.reactjs.org/learn/extracting-state-logic-into-a-reducer).

Their docs also point to the same suggestion I made previously. When any complexity is added a `useReducer` hook would be preferable and the next step.

Another way of saying this would be _difficult to extend_.

### Complexity

Previously you introduced `async` to the problem space. That's an extra level of complexity.

However, my suggestions to steer away from `updaters` and use `reducers` to manage that problem space. 

useState vs useReducer
https://stackblitz.com/edit/react-ts-aost1g?file=App.tsx

useState vs useReducer async
https://stackblitz.com/edit/react-ts-uhtx5n?file=App.tsx

Which also brings up a good question, what exactly is the difference between `useState(updater)` and `useReducer`

*****

I don't know where else to go with this comment thread. Let's book time to talk more in person if there is further discussion needed.

Reply:
After today's Convo we I believe we've narrowed this down to two potential solutions:
1) anonymous function callback
2) usecallback callback

I think we can agree we are not moving forward with the updater.

I'm still in favor of useCallback(2) as it's a more complete example. And more importantly a cached callback will not drive child updates (despite only having one child)

The anonymous function wil cause the child component to update. The child component in this case is a forwarded ref and it will update if any properties change

If it's too much we can update after this PR. But I feel comfortable with this solution 

We can always update later if it really bothers other engineers. After this submitted, I'll be working on other things and won't be invested as much.

If this bricks this PR, I'll find it a little silly and over the top but down to talk more

Reply:
My preference is 1. We've both presented our cases on our preferred usage of `useCallback` and failed to come to an agreement. It's only example code. I won't block the merging of this PR whichever way you go.

Reply:
Thank you @augustjk

Reply:
resolving because it's a mile long